backend/src/apiserver/resource/resource_manager.go

```diff
@@ -229,6 +236,117 @@ func (r *ResourceManager) GetPipelineTemplate(pipelineId string) ([]byte, error)
 	return template, nil
 }
 
+// This only detects default usage of use_gcp_secret sdk method.
+func isGcpSecretVolume(volume apicorev1.Volume) bool {
+	return volume.Secret != nil &&
+		volume.Name == defaultGcpSecretVolumeName &&
+		volume.Secret.SecretName == defaultGcpSecretName &&
+		(volume.Secret.Optional == nil || !*volume.Secret.Optional)
+}
+
+func isGcpSecretVolumeMount(volumeMount apicorev1.VolumeMount) bool {
+	return volumeMount.Name == defaultGcpSecretVolumeName
+}
+
+// user-gcp-sa secret was the default way to authenticate in GCP, but it is no longer needed when
+// we suggest using default authentication methods provided by GKE.
+// Alternatives include: project default service account, node pool default service account and
+// workload identity.
+// Dropping GOOGLE_APPLICATION_CREDENTIALS env when the secret is not configured ensures old
+// pipelines can still be run in new recommended clusters set up as above methods.
+// Reference, old pipeline mounts secret like https://github.com/kubeflow/pipelines/blob/e19efdff76829ee374f400e68cc478ebff53a845/sdk/python/kfp/gcp.py#L46
+//
+// TODO: when all pipeline samples no longer use user-gcp-sa secret, we can remove this usage.
+func (r *ResourceManager) dropUserGcpSaIfNotConfigured(workflow util.Workflow) error {
+	someStepHasGcpSecretVolume := false
+	for _, template := range workflow.Spec.Templates {
+		for _, volume := range template.Volumes {
+			if isGcpSecretVolume(volume) {
+				someStepHasGcpSecretVolume = true
+				break
+			}
+		}
+	}
+	hasGlobalGcpSecretVolume := false
+	for _, volume := range workflow.Spec.Volumes {
+		if isGcpSecretVolume(volume) {
+			hasGlobalGcpSecretVolume = true
+			break
+		}
+	}
+	if !(someStepHasGcpSecretVolume || hasGlobalGcpSecretVolume) {
+		// No steps declared user-gcp-sa secret, no need to drop GOOGLE_APPLICATION_CREDENTIALS.
+		return nil
+	}
+	_, err := r.secretClient.Get(defaultGcpSecretName, v1.GetOptions{})
+	if err == nil {
+		// user-gcp-sa secret is set up correctly, no need to drop GOOGLE_APPLICATION_CREDENTIALS.
+		return nil
+	}
+	if !util.IsNotFound(err) {
+		return errors.Wrap(err, "Failed to verify whether user-gcp-sa secret exists or not because of"+
+			" a runtime error. If you don't need this secret (e.g. when you are using workload identity"+
+			" or default GKE service account), please remove use_gcp_secret usage from your pipeline."+
+			" Otherwise, raise an issue in kubeflow/pipelines Github repo.")
+	}
+
+	glog.Infof("Workflow %s mounts user-gcp-sa secret, but the secret is not present. Dropping GOOGLE_APPLICATION_CREDENTIALS from the workflow and marking gcp-user-sa secret volumes as optional.", workflow.Name)
+	for _, volume := range workflow.Spec.Volumes {
+		if isGcpSecretVolume(volume) {
+			overrideOptional := new(bool)
+			*overrideOptional = true
+			volume.Secret.Optional = overrideOptional
+		}
+	}
```

Comment:
FYI, only remove the ENV Var is enough, but OK here you also do these clean up.

Reply:
ack, got your meaning to have more info for debugging

Reply:
I prefer leaving the secret optional here to give them a clue what happened, just in case they may hit this unexepectedly.